### ISSUE TITLE
No get_as_bulk_array method on multimesh

### ISSUE THREAD

**Issue by puchik** [godotengine/godot#47158]
**Godot version:**

3.2.3 stable

**OS/device including version:**

Windows 10 64bit, Ryzen 3600, RTX 2070s, Vulkan/GLES3/GLES2

**Issue description:**

Not sure if it is a feature proposal or a bug, but seems closer to a bug: There is a `set_as_bulk_array` method, but no `get_as_bulk_array`.

**Steps to reproduce:**
Try to use `get_as_bulk_array()`, which does not exist.

**Minimal reproduction project:**

Not necessary? It's a missing function.

**Comment by akien-mga** [comment]
Could you explain why you want the method/what the use case would be? The existence of a setter doesn't necessarily mean that there should be a getter if the data is only meant as input and not output.

**Comment by puchik** [comment]
Fair enough, I probably should have explained originally!

Personally, I discovered this when trying to sort multimesh instances by closest to the player camera (so if I limited the number of visible instances, only the closest ones are left), but accessing `get_instance_transform` alone (not to mention set_instance_transform) seems to be too slow when handling large amounts of multimeshes (for some reason, it is much slower on a thread, as well... maybe sync issues?), especially when you're doing something with many accesses like an in-place sort. I didn't want to build a new sorted array either because of the large number of objects, and the accesses to build it initially would still be too slow. But if I can fetch all of them at once, do my processing in a thread, then instantly bring them back, it would make it much faster.

I can see other uses for this, too. The documentation talks about `set` being useful for loading data and processing it, but in the same way it could be useful for saving data or mutating already existing multimesh data. Saving could be useful for custom levels, for example. You may also want to assign many new properties to the multimesh that require you to know the original values. For example, you have many rocks instanced as multimeshes, and it begins to rain, and you want to mark some rocks as not having objects above them using their transform so they can become wet.

**Comment by Zylann** [comment]
> accessing get_instance_transform alone (not to mention set_instance_transform) seems to be too slow when handling large amounts of multimeshes 

I stumbled across the same need in my voxel engine. I have a system that spawns instances on top of terrain using multimeshes (grass and small rocks) and these instances must be detached when the user digs holes in the ground where they stand. So I need to iterate them to find which ones must be removed. I would probably not need `get_as_bulk_array` since it would still need decoding, but I've seen transforms are cached in `VisualServer` so it would be useful to just have access to that, instead of doubling my own cache or having to do an invididual call for each.

**Comment by clayjohn** [comment]
I think this could be a harmless addition. So I am not opposed to adding it. That being said, I am not certain it is necessary or advisable to use. 

``set_as_bulk_array`` was added in order to support CPU Particles. We needed a fast way to update all the particle data at once. The use-case is similar to the above, except CPU Particles don't need to request data from the visual server.  CPU Particles maintains its own particle array and then just pushes to the Visual Server when it needs to. This works well with Godot's design which is to try to never return data from the server (see the [Servers Doc](https://docs.godotengine.org/en/stable/tutorials/optimization/using_servers.html#getting-data-from-the-servers)). 

If performance is your main concern, then you should keep a copy of the array yourself. ``get_as_bulk_array`` is going to cause threading stalls and will decrease overall performance. Off the top of my head the only time ``get_as_bulk_array`` would be advisable is in a single-threaded environment where memory is extremely constrained.

**Comment by Zylann** [comment]
> If performance is your main concern, then you should keep a copy of the array yourself. 

That sounds sensible, however why is VisualServer keeping a duplicated cache as well?

**Comment by clayjohn** [comment]
> That sounds sensible, however why is VisualServer keeping a duplicated cache as well?

The typical API for Multimesh is for users to allocate a certain number of instances in advance (this allocates space on the GPU), and then set each instance one at a time using commands (``set_instance_transform`` for example). ``set_as_bulk_array`` was introduced to allow users to set all transforms, color data, and custom data in a single API call to reduce overhead. This API requires an array to be stored in the renderer. Further, users can still request instance data from the API using, for example, ``get_instance_transform``

**Comment by Zylann** [comment]
Hmm... having two copies of the same data CPU-side in such case sounds wasteful, but that's probbaly unavoidable to support use cases Godot wants to provide I suppose

**Comment by puchik** [comment]
> If performance is your main concern, then you should keep a copy of the array yourself. `get_as_bulk_array` is going to cause threading stalls and will decrease overall performance. 

Yeah I was considering only using it in the beginning to get the initial values then mutating and setting it when needed. The threading stalls actually explain why the iteration performance was so much worse than in the main thread.

> Transforms are cached in the same format as *bulk_array in VisualServer so it would be useful to just have access to that

That does sound good, but I'm not sure if that's possible to do with GDScript?

I do have a branch with a `get_as_bulk_array` implementation that makes a PoolArray copy it sends out. I can make a PR for that.

**Comment by clayjohn** [comment]
> Hmm... having two copies of the same data CPU-side in such case sounds wasteful, but that's probably unavoidable to support use cases Godot wants to provide I suppose

I guess, realistically how large is your instance array? We are talking about data in the range of kilobytes for an array of thousands of instances. But yeah, thats why I mentioned above that ``get_as_bulk_array`` may be useful for extremely memory constrained situations. 

> Transforms are cached in the same format as *bulk_array in VisualServer so it would be useful to just have access to that

From a usability standpoint it would be great to just write directly to or read directly from the transform array, but then multithreading is just not possible at all. The architecture of the Visual Server is designed to allow for maximum speed and multithreading opportunities. Sometimes this results in a slightly more verbose API, but the performance benefits are certainly worth it.

**Comment by Zylann** [comment]
>  it would be great to just write directly to or read directly from the transform array, but then multithreading is just not possible at all.

Isn't it when CoW actually comes in handy?
